internal/keepers/service.go

```diff
@@ -89,12 +89,18 @@ func (s *onDemandUpkeepService) CheckUpkeep(ctx context.Context, key types.Upkee
 	if err != nil {
 		return result, err
 	}
-	state, stateCached := s.stateCache.Get(string(id))
 
+	// if an upkeep is in-flight (reported but not confirmed) all checks should
+	// return the InFlight status
+	_, isInFlight := s.inFlightCache.Get(string(id))
+
+	// the cache is a collection of keys (block & id) that map to cached
+	// results. if the same upkeep is checked at a block that has already been
+	// checked, return the cached result
 	result, cached := s.cache.Get(string(key))
 	if cached {
-		if stateCached {
-			result.State = state
+		if isInFlight {
+			result.State = types.InFlight
```

Comment:
I think we can separate out these states.

1. State for checkUpkeep - eligible or not -> depends on geth call
2. State for inFlight - inFlight or not -> depends on plugin execution

Reply:
If we move the inflight ids to a separate data structure and pass that into the service, that would simplify the service a good bit.